xds/internal/balancer/weightedtarget/weightedtarget.go

```diff
@@ -0,0 +1,150 @@
+	"google.golang.org/grpc/internal/hierarchy"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+	"google.golang.org/grpc/xds/internal"
+	"google.golang.org/grpc/xds/internal/balancer/balancergroup"
+)
+
+const weightedTargetName = "weighted_target_experimental"
+
+func init() {
+	balancer.Register(&weightedTargetBB{})
+}
+
+type weightedTargetBB struct{}
+
+func (wt *weightedTargetBB) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	b := &weightedTargetBalancer{}
+	b.logger = grpclog.NewPrefixLogger(loggingPrefix(b))
+	b.bg = balancergroup.New(cc, nil, b.logger)
+	b.bg.Start()
+	b.logger.Infof("Created")
+	return b
+}
+
+func (wt *weightedTargetBB) Name() string {
+	return weightedTargetName
+}
+
+func (wt *weightedTargetBB) ParseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	return parseConfig(c)
+}
+
+type weightedTargetBalancer struct {
+	logger *grpclog.PrefixLogger
+
+	// The keys in BalancerGroup is currently a locality struct, because
+	// balancer group also does load reporting (which is disabled here since
+	// LoadStore isn't set). The keys are all set to Locality{Region: name}
+	// until load reporting is refactor and moved out from balancer group.
+	bg *balancergroup.BalancerGroup
+
+	targets map[string]target
+}
+
+// TODO: remove this and use strings directly as keys for balancer group.
+func makeLocalityFromName(name string) internal.LocalityID {
+	return internal.LocalityID{Region: name}
+}
+
+// UpdateClientConnState takes the new targets in balancer group,
+// creates/deletes sub-balancers and sends them update. Addresses are split into
+// groups based on hierarchy path.
+func (w *weightedTargetBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	newConfig, ok := s.BalancerConfig.(*lbConfig)
+	if !ok {
+		return fmt.Errorf("unexpected balancer config with type: %T", s.BalancerConfig)
+	}
+	addressesSplit := hierarchy.Group(s.ResolverState.Addresses)
+
+	// Remove sub-balancers that are not in the new config.
+	for name := range w.targets {
+		if _, ok := newConfig.Targets[name]; !ok {
+			w.bg.Remove(makeLocalityFromName(name))
+		}
+	}
+
+	// For sub-balancers in the new config
+	// - if it's new. add to balancer group
+	// - if it's old, but has a new weight, update weight in balancer group.
+	//
+	// For all sub-balancers, forward the address/balancer config update.
+	for name, newT := range newConfig.Targets {
+		l := makeLocalityFromName(name)
+
+		oldT, ok := w.targets[name]
+		if !ok {
+			// If this is a new sub-balancer, add it.
+			w.bg.Add(l, newT.Weight, balancer.Get(newT.ChildPolicy.Name))
+		} else if newT.Weight != oldT.Weight {
+			// If this is an existing sub-balancer, update weight if necessary.
+			w.bg.ChangeWeight(l, newT.Weight)
+		}
+
+		// Forwards all the update:
+		// - Addresses are from the map after splitting with hierarchy path.
+		// - Top level service config and attributes are the same
```

Comment:
Nit: Here and in some other comment lines, please terminate with a period.

Reply:
Done